### PR TITLE
Add SoftNPU and P9fs devices.

### PR DESCRIPTION
**Pull request by rcgoodfellow**
This PR adds two new devices to Propolis.

- A P4 ASIC emulator called SoftNPU.
- A Plan 9 filesystem device.

These devices have been in daily use through [Falcon](https://github.com/oxidecomputer/falcon) for quite some time now. This PR is an effort to

- Make these devices easier for folks to consume by being in mainline Propolis.
- Reduce maintenance overhead in keeping the `falcon` branch up to date.

A synopsis of these devices can be found in the main struct comments for [SoftNPU](https://github.com/oxidecomputer/propolis/blob/softnpu-pr/lib/propolis/src/hw/virtio/softnpu.rs#L45-L89) and [PciVirtio9pfs](https://github.com/oxidecomputer/propolis/blob/softnpu-pr/lib/propolis/src/hw/virtio/p9fs.rs#L41-L51).

Today these devices are used exclusively in development environments, and that is likely to remain true indefinitely for SoftNPU. SoftNPU supports loading shared library code given to it by programs running in the guest. This is not something most people would want anywhere near a production environment hypervisor. Due to these characteristics, both of these new devices are only compiled into Propolis if the `falcon` Cargo feature is present.

At the time of writing, this PR depends on a private `p4` repository. That repository will be going public soon. After that happens and after a bit more self-review, I'll move this PR out of the draft state.

### CODE REVIEW

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/softnpu.rs:273]: Given that only the `start` handler is implemented, is it safe to say that falcon guests (or at least those using SoftNPU) are never expected to reboot? (Or for that matter, cleanly halt)

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/softnpu.rs:110]: Given how `data_handles`, the (CPU port) virtio state, and pipeline are all used by the worker threads, does it make sense to keep extra copies here (where they are passed to said workers during spawn), or perhaps just hand off the `Arc<PciVirtioSoftNpuPort>`, into which those workers can reach for the bits they need?

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/softnpu.rs:392]: The virtio abstraction(s) in propolis should probably grow a way to get/set `USED_F_NO_NOTIFY`.  The kernel side of viona does this, indicating to the guest that we do not want/need notifications for newly available descriptors in the RX ring.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/softnpu.rs:606]: We should probably be more graceful here if we cannot access guest memory through the pci state handle.  Such would be the case if the guest happened to reset the device and/or disable bus-mastering.  In that case, this handler can assume there is no free descriptor into which the packet can be placed.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/softnpu.rs:523]: Any particular reason why the virtio refs are being cloned for every packet?

File: lib/propolis/Cargo.toml
Comment (pfmooney) [inline lib/propolis/Cargo.toml:56]: For ease of reading, can we include a comment here mentioning that these are associated w/ falcon too?

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (pfmooney) [inline lib/propolis-client/src/instance_spec/devices.rs:323]: I'll defer to @gjcolombo on the instance spec stuff here.

**Review by pfmooney — COMMENTED** [review]
Thanks for putting this together.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [reply lib/propolis/src/hw/virtio/softnpu.rs:392]: In short: Ignoring notifications to vq0 is the right thing to do here.

File: bin/propolis-server/Cargo.toml
Comment (rcgoodfellow) [reply bin/propolis-server/Cargo.toml:68]: Yeah, I think so. Seems I was getting away with this based on the fact that I was building everything all the time via `cargo build --features falcon` which applied the feature across the cargo projects.

File: lib/propolis-client/src/instance_spec/builder.rs
Comment (gjcolombo) [inline lib/propolis-client/src/instance_spec/builder.rs:179]: This should also call `self.register_pci_device` so that the builder knows the port is reserved (it will then catch attempts to use it again). Same thing for SoftNpuP9.

File: lib/propolis/src/hw/virtio/p9fs.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/p9fs.rs]: SGTM. I've transitioned to a trait object in 1abd8d7.

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (gjcolombo) [inline lib/propolis-client/src/instance_spec/devices.rs:286]: Minor nit: if this VNIC is similar to a `NetworkDevice` VNIC, i.e. it's some sort of host resource that's configured/managed outside of Propolis, then strictly speaking it should be defined over in the backend spec (and then this element can refer to that backend element by name the way a `NetworkDevice` does). I'm not sure if the existing virtio backend is fit for purpose here or if you'd need to define another backend type.

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (gjcolombo) [inline lib/propolis-client/src/instance_spec/devices.rs:326]: I'm assuming P9 and SoftNpu devices aren't meant to live migrate for now? If they aren't, their `Entity` impls should indicate as much (see my comment over in softnpu.rs). If they are, the `can_migrate_devices_from` routine (which I cannot comment on directly even though it is LITERALLY ON THE NEXT LINE come on Github are you kidding me) should be updated to check these elements for compatibility.

If it existed, the doc comment for `can_migrate_devices_from` would explain that it returns `Ok(())` if `self` and `other` describe specs whose members are similar enough that if you initialized instance 1 from `self` and instance 2 from `other`, you could migrate from one to the other without breaking any guest abstractions (e.g. changing something's PCI path).

Again, my guess is that for now it's best for the components themselves just to say that they can't be migrated, but if you do want them to migrate these checks are important to add.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (gjcolombo) [inline lib/propolis/src/hw/virtio/softnpu.rs:268]: Are SoftNpu and/or P9 devices expected to be migratable? If not they should also implement `migrate` and return `Migrator::NonMigratable`. (If they are then you probably need some instance spec changes; left a comment over in that module describing those.)

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (gjcolombo) [inline lib/propolis-client/src/instance_spec/devices.rs:305]: Nit: doc comments for these fields

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (gjcolombo) [reply lib/propolis-client/src/instance_spec/devices.rs:323]: Thanks! I've left a few comments.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/softnpu.rs:273]: An interesting question. For the primary thing that SoftNPU currently emulates, a sidecar switch - the sidecar will indeed keep on going if the attached scrimlet reboots, halts, etc. So in that sense, I think we have the desired behavior here, where the guest (scrimlet) reboots and comes back to find the sidecar not having gone down with it.

If we start using SoftNpu for emulating something like a programmable NIC, that would be a different story.

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (rcgoodfellow) [reply lib/propolis-client/src/instance_spec/devices.rs:286]: I think this would be a different backend type than virtio. Virtio plugs into the VNIC with a viona data path, SoftNpu ports use DLPI.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/softnpu.rs]: Nope

File: lib/propolis/Cargo.toml
Comment (rcgoodfellow) [reply lib/propolis/Cargo.toml:57]: Done

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/softnpu.rs]: Now checking for none, kicking out a warning and returning.

File: lib/propolis-client/src/instance_spec/builder.rs
Comment (rcgoodfellow) [reply lib/propolis-client/src/instance_spec/builder.rs:180]: Done

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (rcgoodfellow) [reply lib/propolis-client/src/instance_spec/devices.rs]: This field and the machinery around it have been updated to used a registered backend device.

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (rcgoodfellow) [reply lib/propolis-client/src/instance_spec/devices.rs:334]: Yeah, we don't need to worry about migration for these devices now. If P9fs support graduates from a Falcon feature to a product feature, we can worry about migration then. For now, Falcon does not migrate things ;)

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/softnpu.rs]: This is all tucked away in the `PciVirtioSoftnpuPort` now, copies have been removed from `SoftNpu`.

**Comment by rcgoodfellow** [comment]
Thanks for the reviews @pfmooney and @gjcolombo!

I've made updates based on the feedback.

File: lib/propolis-client/src/instance_spec/devices.rs
Comment (gjcolombo) [reply lib/propolis-client/src/instance_spec/devices.rs:334]: Works for me!

**Review by gjcolombo — APPROVED** [review]
LGTM for the instance spec and instance lifecycle pieces.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/softnpu.rs:315]: I would probably just silently return like the `pop_avail` check below.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [reply lib/propolis/src/hw/virtio/softnpu.rs:262]: To be clear: I'm not suggesting that we should reset the emulated switch state, but as-is, we're ignoring a bunch of expected state management expected of an emulated (PCI) device if the guest undergoes reboot.

If you really do not expect a falcon guest (at least one with a SoftNPU device attached) to undergo reboot, then this is probably fine, but as-is, it is breaking some of the rules/expectations around how device emulation acts when the instance is walking through expected state transitions for a reboot or halt.

File: lib/propolis/src/hw/virtio/p9fs.rs
Comment (pfmooney) [inline lib/propolis/src/hw/virtio/p9fs.rs:92]: I should have caught this earlier: this could probably be reduced to a `as_bytes()` call on the `&str`, rather than a buffer copying situation.  In both cases, some bounds checking is likely needed to avoid potential overflow.

File: bin/propolis-server/src/lib/server.rs
Comment (pfmooney) [inline bin/propolis-server/src/lib/server.rs:276]: Lingering question for @gjcolombo: Is there a future where we might want to attach serial ports to things other than a user console?  In such a case, where COM port attachment is described in the instance spec, we could probably avoid its omission here (in the falcon case), and just spell out that it's being used by the SoftNPU?

Comment (pfmooney) [reply bin/propolis-server/src/lib/server.rs:276]: I also wonder if there's a device id we could "borrow" for a PCI-attached 16650 chip, which the SoftNPU could just attach at slot X function 1, where that exposed topology implied the connection between the UART and the NPU stuff.

Comment (gjcolombo) [reply bin/propolis-server/src/lib/server.rs:276]: > Lingering question for @gjcolombo: Is there a future where we might want to attach serial ports to things other than a user console? In such a case, where COM port attachment is described in the instance spec, we could probably avoid its omission here (in the falcon case), and just spell out that it's being used by the SoftNPU?

I'd model that as a device/backend split: the device handles all the UART register emulation, and the bytes it receives/transmits are sent to/obtained from some backend, which could be a human-readable I/O channel of some kind, a SoftNPU device, etc. I'd have to look at the serial port code again (it's been a while) to see how big of a refactoring that would be.

This will also be sort of relevant when #231 lands, because it lets callers like Omicron and PHD pass instance specs directly to Propolis, and they won't take this path. But I don't think those consumers plan to use Falcon builds, if they do they probably won't attach a SoftNPU device, and if they do we can catch/resolve the conflict in Propolis (today I'm pretty sure it'll panic because two devices will try to map the same part of PIO space, but we can make that cleaner by checking the instance spec for a conflict). So I don't think we need to refactor anything right away or change anything in this PR (though I might want to touch up #231 a bit after this lands).

I've filed #259 to track the broader "serial ports should have devices and backends" idea.

File: lib/propolis/src/hw/virtio/p9fs.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/p9fs.rs]: When I change this to 
```
P9fsReg::Tag => {
    ro.write_bytes(self.handler.target().as_bytes());
    ro.fill(0);
}
```

The result is
```
thread 'vcpu-0' panicked at 'source slice length (14) does not match destination slice length (256)', lib/propolis/src/common.rs:156:29
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::slice::<impl [T]>::copy_from_slice::len_mismatch_fail
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/slice/mod.rs:3239:13
   3: propolis::common::ReadOp::write_bytes
   4: propolis::util::regmap::RegMap<ID>::process
   5: <propolis::hw::virtio::p9fs::PciVirtio9pfs as propolis::hw::virtio::VirtioDevice>::cfg_rw
   6: propolis::util::regmap::RegMap<ID>::process
   7: propolis::hw::pci::device::<impl propolis::hw::pci::Endpoint for D>::bar_rw
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I've added a check to the current code to avoid overruning the buffer for tag strings longer than 256 bytes.

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/softnpu.rs]: done

File: lib/propolis/src/hw/virtio/softnpu.rs
Comment (pfmooney) [reply lib/propolis/src/hw/virtio/softnpu.rs:262]: I'd overlooked the fact that the `PciVirtioSoftNpuPort` device had wired up the appropriate events for ensuring a PciVirtio reset when the guest itself is rebooted.  While there are some improvements that could be made around how the dlpi recv thread(s) are being managed (they'll likely prevent clean propolis exit on instance halt), that is something which could be addressed in a follow-up.  Clearly the tools in place for using Falcon today already take that into account.

Comment (rcgoodfellow) [reply lib/propolis/src/hw/virtio/softnpu.rs:262]: I've added an issue to track this 
- https://github.com/oxidecomputer/propolis/issues/261

**Review by pfmooney — APPROVED** [review]
Thanks for putting this together.

**Pull request merged**
